cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,455 @@
+
+Luckily, we have something that already has all the information it needs for all of those things - the cilium-agent itself. And also luckily, when some other folk worked on ideas about integrating SPIFFE into Cilium earlier, they also did a bunch of work with SPIFFE and SPIRE to create a Delegated Identity API and implemented it in the SPIRE server. What this means is that we can give all the Cilium agents a common SPIFFE identity, register that identity with the SPIRE server, and then grant those identities the permission to be delegates and watch for identities on behalf of other workloads. Those identities will be sent to the Cilium Agent as soon as they are created on the SPIRE server, as long as the SPIFFE selectors (in our case, the label selectors) are in the set being watched by the Cilium Agent.
+
+This diagram shows the way the API flows work. It’s assumed before we start that SPIFFE identities for the source and destination have already been created in the SPIRE server.
+
+![SPIFFE and SPIRE Cilium Agent using Delegated Identity API](./images/SPIFFE-SPIRE-cilium-agent-delegated.png)
+
+Then, when required (whether we do this at Endpoint creation or later is still TBD):
+1. All cilium-agents running a workload mentioned in an auth-enabled policy connect to their local spire-agent, and watch a set of SPIFFE labels that includes the labels for the relevant SVID. The cilium-agent identity is allowed to do Delegated Identity requests to the Workload API (WL API), so this is allowed.
+1. The spire-agent watches SVIDs on the SPIRE server, sees any updates to all relevant SVIDs, and passes them back to the cilium-agent.
+1. The cilium-agent then figures out where the request is coming from, and connects to the cilium-agent on the source node to perform the mTLS handshake. Because this  a mutual TLS, if it succeeds, then the workloads are authenticated.
+The destination node will also record that the workload is authenticated for reverse traffic.
+1. The mTLS handlers in each cilium-agent then pass the auth success to their local dataplane, by telling it that identity A is authenticated to identity B for some period of time (the lifetime of the certificates in the mTLS handshake).
+The mTLS handlers will also need to indicate the remote node, along with the auth type.
+For that lifetime, traffic between Identity A and Identity B will flow without further checks.
+
+#### SPIFFE installation steps and flow
+
+1. Cluster gets created, without CNI, per usual Cilium install process. 
+1. Cilium installed, with mTLS enabled, includes SPIRE Server and Cilium Agent SPIFFE Identity
```

Comment:
I agree that an imperative form will be clearer for this set, I'll work on changing it.

To answer your questions though:
> Does this mean that the SPIRE server is part of the Cilium helm charts?

For the initial release, we will _at least_ be supporting installing SPIRE into the cluster for you as part of the Cilium Helm charts, yes. Having the feature be able to be used in a "turn it on, everything works" way is an important part of the initial release.

> What about users who want to bring their own?

We absolutely will allow this at some point; I'm hoping that this will also be available in 1.14, as well as the auto-install.

> What is the lifecycle of the SPIRE server in relation to Cilium? Do you have to upgrade these at the same time?

We haven't spent a lot of time thinking about lifecycle yet, but I don't _think_ that they'll need to be upgraded at the same time. I anticipate the SPIRE server and spire agent will be their own StatefulSet and Daemonset respectively, and will be able to be upgraded independently of the agent if required (no identities will be issued while they are down).

> Is there a strict relationship between Cilium version and SPIRE version for compatibility?

There's nothing to indicate that at this point; the API extension we are using (DelegatedIdentity) has been available in both SPIFFE and SPIRE for some time.

> Is the "Cilium Agent SPIFFE" identity something that comes with Cilium Helm charts or a configuration that must be independently installed into SPIRE?
It is a configuration that needs to be installed into SPIRE, but we are planning to have this part of the Helm install process for the automagic installation. For bring-your-own in the future, we'll need to write out the steps and identities that Cilium and the Operator expect.

I'll try and fold these answers in to the document tomorrow.

Reply:
> > What is the lifecycle of the SPIRE server in relation to Cilium? Do you have to upgrade these at the same time?
> 
> We haven't spent a lot of time thinking about lifecycle yet, but I don't think that they'll need to be upgraded at the same time. I anticipate the SPIRE server and spire agent will be their own StatefulSet and Daemonset respectively, and will be able to be upgraded independently of the agent if required (no identities will be issued while they are down).

At Isovalent we've had some pain points around including the DaemonSets etc. for other components in the same Helm chart, because you can't always control which pieces to upgrade. If we make a minor tweak in the SPIRE config in the new Cilium Helm chart, this can mean that a `helm upgrade ...` to the new version will cause redeployment of the SPIRE server. This is the main reason I raise this, because even though I agree it seems simpler to have one less `helm install` command, later on this can cause lifecycle upgrade headaches.

Reply:
Regarding that last comment, SPIRE does have their own Helm chart, so we might be able to do something with sub-charts. Otherwise, the fallback will need to be "install SPIRE in the cluster per these instructions". That has the advantage that we can also write instructions for "bring your own SPIRE server that needs to have the following properties" later, which will allow bring-your-own-SPIRE-server.